mettle/extensions/lua/lua.c

```diff
@@ -0,0 +1,61 @@
+#include "extension.h"
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lualib.h>
+
+#include "lua.h"
+
+
+/*
+ * Execute the code supplied (via TLV packet) string
+ */
+static struct tlv_packet *request_execute_code(struct tlv_handler_ctx *ctx)
+{
+	lua_State *lua;
+
+	lua = luaL_newstate();
+	luaL_openlibs(lua);
+
+	int tlv_result = TLV_RESULT_FAILURE; // By default
+	struct tlv_packet *r = tlv_packet_response(ctx);
+
+	char *execute_me = tlv_packet_get_str(ctx->req, TLV_TYPE_LUA_CODE);
+	if (luaL_dostring(lua, execute_me) == 0) {
```

Comment:
For some reason this call never return when executed in mettle's context, thus executing valid code times out, any hints?

Reply:
Are you certain you are using Lua API correctly? Have you tried a simple driver program that has the same includes and performs the same steps using Lua from a `char *`? Have you output the value of `execute_me` and confirmed it to be correct?

Reply:
Hello,

Yes, the code has been tested in a simple program (can send you an example if you want) which worked perfectly, also checked the `luaL_dostring` prototype and I am using the right types of variables.

Using the file at https://gist.github.com/DeveloppSoft/336e6f3086d1e691c27567da5ac1c0c5 along with invocating mettle as `./mettle -d 100 -u tcp://127.0.0.1:4444` doesn't print anything (not even a carriage return in the console).

Managed to log some stuff by using a file instead (https://gist.github.com/DeveloppSoft/2e1aea80be26d8241c47c60b7685d2dc):
Msf side:
```
meterpreter > load lua
Loading extension lua...WARNING: Local file /home/eliott/Documents/Documents/metasploit-framework/data/mettle/x86_64-linux-musl/bin/lua is being used
Success.
meterpreter > lua_dostring "hj"
[-] lua_dostring: Operation failed: 1
meterpreter > lua_dostring "a=1"
[-] Error running command lua_dostring: Rex::TimeoutError Operation timed out.
meterpreter > 
```

Mettle side:
```
eliott@pop-os:~/Documents/Documents/mettle/build/x86_64-linux-musl/bin$ ./mettle -d 100 -u tcp://127.0.0.1:4444
[04-30-2018 19:19:48.785s] [process.c:392] child pid 14816 started
[04-30-2018 19:19:48.785s] [process.c:428] IO started on fds 7 9
```

Finally the log file:
```
eliott@pop-os:~/Documents/Documents/mettle/build/x86_64-linux-musl/bin$ cat LOGMEHERE.txt 
a=1
```

So I think we can conclude the code is received the right way, it also seems that sending a command to print something from lua such as `lua_dostring "print 'hello'"` produces a segmentation fault, haven't yet confirmed if this comes from the extension (most probably) or mettle's core (don't think so).

Even if this is a pre GSOC task which I choose to make sure I understand mettle's internal, I would still like to manage to get it done the right way, thus will continue to dive in :wink:.

Reply:
Just a guess, but since the extension is using stdin/stdout for communication surely if the lua_dostring prints something to stdout something will crash?

Reply:
Sadly yes, I guess I could fork to a new process or something to mitigate this ?